### PR TITLE
FISH-663 MicroProfile OpenAPI 2.0 (MP 4.0)

### PR DESCRIPTION
**Pull request by MattGill98**
## Description

This is an improvement, upgrading our implementation of MicroProfile OpenAPI to version 2.0

Changes include:
- The @SchemaProperty annotation has been added to allow the properties for a schema to be defined inline
- The @RequestBodySchema annotation has been added to provide a shorthand mechanism to specify the schema for a request body
- The @APIResponseSchema annotation has been added to provide a shorthand mechanism to specify the schema for a response body
- The `mp.openapi.schema.*` MicroProfile Config property has been added to allow the schema for a specific class to be specified. This property would typically be used in cases where the application developer does not have access to the source code of a class
- Getter methods on model interfaces that return a list or map now return an immutable copy of that collection
- Setter methods on model interfaces that take a list or a map as a parameter MUST not use the list/map instance directly

The TCK has been upgraded in various places, meaning that not all of the changes here are strictly due to new behaviour.

## Important Information
HK2 has received a patch to now parse exception types when creating MethodModels. This was needed as exception mappers are now supported in OpenAPI (they may have been expected before, but not implemented).
Patched source: https://github.com/payara/patched-src-hk2/pull/19
Patched artifact: https://github.com/payara/Payara_PatchedProjects/pull/352

### New tests

TCK Upgrade: https://github.com/payara/MicroProfile-TCK-Runners/pull/132

### Testing Performed

Ran the new TCK. OpenAPI is reasonably isolated, so there shouldn't bean affect on many other tests.

## Notes for Reviewers

I've disabled a TCK test and raised an upstream issue for it. I don't believe it's intended, and it shouldn't affect or compatibility.

This PR also fixes https://github.com/payara/Payara/issues/4955 (FISH-760)

### CODE REVIEW

**Comment by Pandrex247** [comment]
```
[ERROR] /C:/Users/AndrewPielage/Git/Payara/appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/processor/ApplicationProcessor.java:[1151,43] cannot find symbol
[ERROR]   symbol:   method getExceptionTypes()
[ERROR]   location: variable method of type org.glassfish.hk2.classmodel.reflect.MethodModel
```
Sad panda

**Comment by MattGill98** [comment]
Did you build the entire server? It builds locally for me

**Review by Pandrex247 — APPROVED** [review]
Haven't gone through the code with a fine-toothed comb, but passes the OpenAPI TCK and seems to work.

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/OpenAPIImpl.java [hunk 1]
```diff
@@ -130,76 +131,95 @@ public void setExternalDocs(ExternalDocumentation externalDocs) {
...
+                    if (server.getUrl().equals(existingServer.getUrl())) {
+                        ModelUtils.merge(server, existingServer, true);
+                        return this;
+                    }
                 }
             }
+    
+            // If a server with the same URL doesn't exist, create it
+            servers.add(server);
         }
```
Comment (MarkWareham) [inline appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/OpenAPIImpl.java:160]:
```suggestion
        if (server == null || server.getUrl() == null)) {
            return this;
        }
       
        if (servers == null) {
            servers = createList();
        }
           
         for (Server existingServer : getServers()) {
              // If a server with the same URL is found, merge them
              if (server.getUrl().equals(existingServer.getUrl())) {
                  ModelUtils.merge(server, existingServer, true);
                  return this;
              }
          }
           
            // If a server with the same URL doesn't exist, create it
            servers.add(server);
```
reduces complexity

**Review by MarkWareham — COMMENTED** [review]
Partial review as Andrew got there before me.

No required changes, but some cleanup suggestions. There's a lot of superfluous null checking going on.

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/ComponentsImpl.java [hunk 2]
```diff
@@ -305,11 +354,13 @@ public static void merge(Components from, Components to,
         }
         // Handle @Schema
         if (from.getSchemas()!= null) {
-            for (String schemaName : from.getSchemas().keySet()) {
+            for (Entry<String, Schema> fromEntry : from.getSchemas().entrySet()) {
```
[the remark this answers is not in the document]
Comment (MattGill98) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/ComponentsImpl.java:357]: This method merged quite a lot of properties, and even if one of them exists it needs merging. I'd argue in favour of the current code

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/parameters/RequestBodyImpl.java [hunk 3]
```diff
@@ -55,19 +57,24 @@
...
         String ref = annotation.getValue("ref", String.class);
         if (ref != null && !ref.isEmpty()) {
             from.setRef(ref);
         }
-        extractAnnotations(annotation, context, "content", ContentImpl::createInstance, from.getContents());
+
+        final List<ContentImpl> contents = createList();
+        extractAnnotations(annotation, context, "content", ContentImpl::createInstance, contents::add);
+        for (ContentImpl content : contents) {
+            ContentImpl.merge(content, from.content, true, context);
```
Comment (jGauravGupta) [inline appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/parameters/RequestBodyImpl.java:75]: As per the existing design of OpenAPI Impl, `merge` operations are invoked after the creation of the instances.
````
    public void visitOpenAPI(AnnotationModel definition, AnnotatedElement element, ApiContext context) {
        OpenAPIImpl.merge(OpenAPIImpl.createInstance(definition, context), context.getApi(), true, context);
    }
````

Comment (MattGill98) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/parameters/RequestBodyImpl.java:75]: What do you mean?

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/ExtensibleImpl.java [hunk 4]
```diff
@@ -58,25 +60,30 @@
...
     public Map<String, Object> getExtensions() {
-        return extensions;
+        return readOnlyView(extensions);
     }
 
     @SuppressWarnings("unchecked")
     @Override
     public T addExtension(String name, Object value) {
         if (value != null) {
+            if (extensions == null) {
```
[the remark this answers is not in the document]
Comment (MattGill98) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/ExtensibleImpl.java:74]: I'd missed a method in my large collection refactor. I've fixed that now, so the extensions can now be null. The TCK asserts than a handful of these collections are still nullable

File: appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/parameters/RequestBodyImpl.java (hunk 3 above)
Comment (jGauravGupta) [reply appserver/payara-appserver-modules/microprofile/openapi/src/main/java/fish/payara/microprofile/openapi/impl/model/parameters/RequestBodyImpl.java:75]: Discussed over impl design in the DM. PR LGTM.

**Review by jGauravGupta — APPROVED** [review]

**Comment by MattGill98** [comment]
jenkins test please

**Pull request merged**